### DISCUSSION TITLE
I wonder how vscode / pylance understands the signature of __init__ function

### DISCUSSION THREAD

**Discussion by ghost** [pydantic/pydantic#6642] [Question] [answered]
I wonder how vscode / pylance understands the signature of __init__ function ?

How does Pydantic make the magic happen?

**Comment by ImogenBits** [comment, marked as answer]
Since dataclasses were widely used type checkers added support for their auto generated init (and other) functions. Some popular libraries (such as pydantic or sqlalchemy) also created similar functionality of auto generating class behaviour based on attribute types. To support those people had to rely on things like mypy plugins or just having everything be annotated with **kwargs. Then [this PEP](https://peps.python.org/pep-0681/) introduced the concept of a dataclass transform, basically user (or, more commonly, library) code can be annotated to tell type checkers that it does similar things to dataclasses. Type checkers now support this feature, i.e. your type checker doesn't actually know what Pydantic is doing to a class, it just knows that it looks something like a dataclass and then treats it as such.

**Reply by samuelcolvin** [reply]
Exactly, dataclass transforms are the answer.

They're not perfect, but they're a good start.
